DevEdu.API/DevEdu.Business/Services/MaterialService.cs

```diff
@@ -10,55 +13,107 @@ public class MaterialService : IMaterialService
+            ITagValidationHelper tagValidationHelper,
+            ICourseValidationHelper courseValidationHelper,
+            IMaterialValidationHelper materilaValidationHelper,
+            IUserValidationHelper useraValidationHelper)
         {
             _materialRepository = materialRepository;
             _courseRepository = courseRepository;
             _groupRepository = groupRepository;
-            _materialValidationHelper = materialValidationHelper;
+            _groupValidationHelper = groupValidationHelper;
             _tagValidationHelper = tagValidationHelper;
+            _courseValidationHelper = courseValidationHelper;
+            _materilaValidationHelper = materilaValidationHelper;
+            _useraValidationHelper = useraValidationHelper;
         }
 
-        public List<MaterialDto> GetAllMaterials() => _materialRepository.GetAllMaterials();
-
-        public MaterialDto GetMaterialById(int id) => _materialRepository.GetMaterialById(id);
+        public List<MaterialDto> GetAllMaterials(UserIdentityInfo user)
+        {
+            var allMaterials = _materialRepository.GetAllMaterials();
+            if (!(user.IsAdmin() || user.IsMethodist()))
+            {
+                return _materilaValidationHelper.GetMaterialsAllowedToUser(allMaterials, user.UserId);
+            }
+            return allMaterials;
+        }
 
         public MaterialDto GetMaterialByIdWithCoursesAndGroups(int id)
         {
-            var dto = _materialRepository.GetMaterialById(id);
+            var dto = _materilaValidationHelper.GetMaterialByIdAndThrowIfNotFound(id);
             dto.Courses = _courseRepository.GetCoursesByMaterialId(id);
             dto.Groups = _groupRepository.GetGroupsByMaterialId(id);
             return dto;
         }
 
-        public int AddMaterial(MaterialDto dto)
+        public MaterialDto GetMaterialByIdWithTags(int id, UserIdentityInfo user)
+        {
+            var dto = _materilaValidationHelper.GetMaterialByIdAndThrowIfNotFound(id);
+            if (!(user.IsAdmin() || user.IsMethodist()))
+            {
+                _materilaValidationHelper.CheckUserAccessToMaterialForGetById(user.UserId, dto);
+            }
+            return dto;
+        }
+
+        public int AddMaterialWithGroups(MaterialDto dto, List<int> tags, List<int> groups, UserIdentityInfo user)
         {
+            _materilaValidationHelper.CheckPassedValuesAreUnique(groups, nameof(groups));
+            foreach (int group in groups)
+            {
+                _groupValidationHelper.CheckGroupExistence(group);
+                if (user.IsTeacher())
+                {
+                    _useraValidationHelper.CheckAuthorizationUserToGroup(group, user.UserId, Role.Teacher);
+                }
+                else 
+                {
+                    _useraValidationHelper.CheckAuthorizationUserToGroup(group, user.UserId, Role.Tutor);
+                }
+            }
```

Comment:
it looks pretty much ugly =)
please use the ForEach method as you've done it before
and also remove this if..else and use ternary operator here